app/controllers/agentAppointment/AgentRelationshipController.scala

```diff
@@ -193,38 +215,72 @@ class AgentRelationshipController @Inject()(
             BadRequest(
               agentToManageMultipleProperties(
                 errors,
-                getAgentName(request)
+                request.agentDetails.name
               )))
-        }, {
-          case ChooseFromList => joinOldJourney(request.sessionData.agentCode)
-          case success @ (propertyrepresentation.All | propertyrepresentation.None) => {
-            sessionRepo.saveOrUpdate(data = request.sessionData.copy(managingProperty = Some(success.name)))
-            Future.successful(Redirect(controllers.agentAppointment.routes.AgentRelationshipController.checkAnswers()))
+        },
+        success => {
+          for {
+            selectedAgentOpt <- sessionRepo.get[SelectedAgent]
+            selectedAgent = selectedAgentOpt.getOrElse(throw NoAgentSavedException("no agent saved"))
+            _ <- sessionRepo.saveOrUpdate(ManagingProperty(selectedAgent, success.name))
+          } yield {
+            success match {
+              case ChooseFromList => joinOldJourney(selectedAgent.agentCode)
+              case propertyrepresentation.All | propertyrepresentation.None =>
+                Redirect(controllers.agentAppointment.routes.AgentRelationshipController.checkAnswers())
+            }
           }
         }
       )
   }
 
-  def checkAnswers(): Action[AnyContent] = authenticated.andThen(withAppointAgentSession).async { implicit request =>
-    Future.successful(Ok(checkYourAnswers(request.sessionData)))
+  def checkAnswers(): Action[AnyContent] = authenticated.andThen(withAppointAgentSession) { implicit request =>
+    PartialFunction
+      .condOpt(request.sessionData) {
+        case data: ManagingProperty =>
+          Ok(checkYourAnswers(submitAppointAgentRequest, data))
+      }
+      .getOrElse(NotFound(errorHandler.notFoundTemplate))
   }
 
-  private def getAgentName(request: AppointAgentSessionRequest[AnyContent]) =
-    request.sessionData.agentOrganisationName.getOrElse(throw new IllegalStateException("no agent name stored"))
+  def appointAgent(): Action[AnyContent] = authenticated.andThen(withAppointAgentSession).async { implicit request =>
+    submitAppointAgentRequest.bindFromRequest.fold(
+      errors => {
+        PartialFunction
+          .condOpt(request.sessionData) {
+            case data: ManagingProperty =>
```

Comment:
I'm a bit confused about the PartialFunction.condOpt
Can you not just match on request.sessionData?

Reply:
IMHO, it would have the same effect. PartialFunction.condOpt was suggested by Dan.

Reply:
Chatted with @DanielConnelly and decided changing this back to a straight match on request.sessionData would be better.